contracts/PangolinExchange.sol

```diff
@@ -0,0 +1,166 @@
+      require(success, "Refund failed");
+    }
+  }
+
+
+  function transferBack(bytes32 _asset) external override {
+    address tokenAddress = supportedAssets.getAssetAddress(_asset);
+    IERC20 token = IERC20(tokenAddress);
+    token.transfer(msg.sender, token.balanceOf(address(this)));
+  }
+
+
+  /**
+   * Buys selected ERC20 token with AVAX using the Pangolin DEX
+   * Refunds unused AVAX to the msg.sender
+   * @dev _token ERC20 token's address
+   * @dev _exactERC20AmountOut amount of the ERC20 token to be bought
+  **/
+  function buyAsset(bytes32 _token, uint256 _exactERC20AmountOut) payable external override RefundRemainder {
+    require(_exactERC20AmountOut > 0, "Amount of tokens to buy has to be greater than 0");
+
+    address tokenAddress = supportedAssets.getAssetAddress(_token);
+
+    uint256 amountIn = getEstimatedAVAXForERC20Token(_exactERC20AmountOut, tokenAddress);
+
+    require(msg.value >= amountIn, "Not enough funds provided");
+
+    pangolinRouter.swapAVAXForExactTokens{value : msg.value}(_exactERC20AmountOut, getPathForAVAXtoToken(tokenAddress), msg.sender, block.timestamp);
+
+    emit TokenPurchase(msg.sender, _exactERC20AmountOut, block.timestamp);
+  }
+
+
+  /**
+   * Sells selected ERC20 token for AVAX
+   * @dev _token ERC20 token's address
+   * @dev _exactERC20AmountIn amount of the ERC20 token to be sold
+   * @dev _minAvaxAmountOut minimum amount of the AVAX token to be bought
+  **/
+  function sellAsset(bytes32 _token, uint256 _exactERC20AmountIn, uint256 _minAvaxAmountOut) external override RefundRemainder {
+    require(_exactERC20AmountIn > 0, "Amount of tokens to sell has to be greater than 0");
+    address tokenAddress = supportedAssets.getAssetAddress(_token);
+
+    IERC20 token = IERC20(tokenAddress);
+    token.approve(address(pangolinRouter), _exactERC20AmountIn);
+    pangolinRouter.swapExactTokensForAVAX(_exactERC20AmountIn, _minAvaxAmountOut, getPathForTokenToAVAX(tokenAddress), msg.sender, block.timestamp);
+
+    emit TokenSell(msg.sender, _exactERC20AmountIn, block.timestamp);
+  }
+
+  /* ========== RECEIVE AVAX FUNCTION ========== */
+  receive() external payable {  }
+
+
+  /* ========== VIEW FUNCTIONS ========== */
+
+  /**
+    * Returns the current balance of the asset held by a given user
+    * @dev _asset the code of an asset
+    * @dev _user the address of queried user
+  **/
+  function getBalance(address _user, bytes32 _asset) external override view returns(uint256) {
+    IERC20 token = IERC20(supportedAssets.getAssetAddress(_asset));
+    return token.balanceOf(_user);
+  }
+
+
+  /**
+     * Returns the minimum token amount that is required to be sold to receive _exactAmountOut of AVAX.
+  **/
+  function getMinimumERC20TokenAmountForExactAVAX(uint256 _exactAmountOut, address _token) public view override returns (uint256) {
+    address[] memory path = getPathForTokenToAVAX(_token);
+
+    return pangolinRouter.getAmountsIn(_exactAmountOut, path)[0];
```

Comment:
Pls check if it might throw an exception

Reply:
@kamilovsky01, could you do it? The biggest risk for us is blocking the liquidation by manipulating single asset.

Reply:
From the perspective of what could go wrong on our end:
According to https://github.com/Uniswap/v2-periphery/blob/master/contracts/libraries/UniswapV2Library.sol this would only fail if we pass the `amountOut` parameter equal to `0` but this is not possible because the `targetAvaxAmount` is calculated as `totalRepayAmount - address(this).balance` and it is only executed inside of a block `if (address(this).balance < (totalRepayAmount)) {}` and later in the for loop we have this statement `if (address(this).balance >= totalRepayAmount) {
          break;
        }`
        
From the perspective of what could go on the Exchange end:
This could revert with an error in a case when:
`require(reserveIn > 0 && reserveOut > 0, 'UniswapV2Library: INSUFFICIENT_LIQUIDITY');`

Do you think we should worry about it though? @jakub-wojciechowski

Reply:
We could validate 0, but it also mean than a single empty pangolinPool could break the liquidation logic. Let's try to catch it in to prevent breaking the liquidation loop.

BTW, I think we should also add a method to allow withdrawing any ERC20 token as long as the loan contract remain solvent (It could protect us against a scenario when token is no longer traded on Pangolin). Could you do it @kamilovsky01, it should be quick.